src/sync_device.rs

```diff
@@ -0,0 +1,707 @@
+            let vals = get_state(&$self.state);
+            let old_vals = get_state(&$self.prev_state);
+            let get_value: fn($state, $typ) -> _ = $get_value;
+            for typ in types_to_check.iter() {
+                let prev = get_value(old_vals, typ);
+                let value = get_value(vals, typ);
+                if prev != value {
+                    let ev = InputEvent(libc::input_event {
+                        time: $time,
+                        type_: EventType::$evtype.0,
+                        code: typ.0,
+                        value: value as _,
+                    });
+                    let state = SyncState::$var {
+                        time: $time,
+                        start: $typ(typ.0 + 1),
+                    };
+                    return Some((ev, state));
+                }
+            }
+        }
+    };
+}
+impl Device {
+    fn compensate_keys(&self, time: libc::timeval, start: Key) -> Option<(InputEvent, SyncState)> {
+        try_compensate!(
+            self,
+            time,
+            start,
+            Key,
+            KEY,
+            Keys,
+            supported_keys,
+            AttributeSet<Key>,
+            |st| st.key_vals().unwrap(),
+            |vals, key| vals.contains(key)
+        );
+        self.compensate_absolutes(time, AbsoluteAxisType(0))
+    }
+
+    fn compensate_absolutes(
+        &self,
+        time: libc::timeval,
+        start: AbsoluteAxisType,
+    ) -> Option<(InputEvent, SyncState)> {
+        try_compensate!(
+            self,
+            time,
+            start,
+            AbsoluteAxisType,
+            ABSOLUTE,
+            Absolutes,
+            supported_absolute_axes,
+            &[libc::input_absinfo],
+            |st| st.abs_vals().unwrap(),
+            |vals, abs| vals[abs.0 as usize].value
+        );
+        self.compensate_switches(time, SwitchType(0))
+    }
+
+    fn compensate_switches(
+        &self,
+        time: libc::timeval,
+        start: SwitchType,
+    ) -> Option<(InputEvent, SyncState)> {
+        try_compensate!(
+            self,
+            time,
+            start,
+            SwitchType,
+            SWITCH,
+            Switches,
+            supported_switches,
+            AttributeSet<SwitchType>,
+            |st| st.switch_vals().unwrap(),
+            |vals, sw| vals.contains(sw)
+        );
+        self.compensate_leds(time, LedType(0))
+    }
+
+    fn compensate_leds(
+        &self,
+        time: libc::timeval,
+        start: LedType,
+    ) -> Option<(InputEvent, SyncState)> {
+        try_compensate!(
+            self,
+            time,
+            start,
+            LedType,
+            LED,
+            Leds,
+            supported_leds,
+            AttributeSet<LedType>,
+            |st| st.led_vals().unwrap(),
+            |vals, led| vals.contains(led)
+        );
+        None
+    }
+}
+
+struct FetchEventsSynced<'a> {
```

Comment:
This also needs some comments/rustdoc, even if it's not publicly visible. A short summary of the approach we're trying to take will help us maintain this later.